crates/spin-wasip3-http/src/lib.rs

```diff
@@ -0,0 +1,526 @@
+use wasip3_http_ext::helpers::{
+    header_map_to_wasi, method_from_wasi, method_to_wasi, scheme_from_wasi, scheme_to_wasi,
+    to_internal_error_code,
+};
+use wasip3_http_ext::RequestOptionsExtension;
+use wasip3_http_ext::{IncomingRequestBody, IncomingResponseBody};
+
+/// A alias for [`std::result::Result`] that uses [`Error`] as the default error type.
+///
+/// This allows functions throughout the crate to return `Result<T>`
+/// instead of writing out `Result<T, Error>` explicitly.
+pub type Result<T, E = Error> = ::std::result::Result<T, E>;
+
+type HttpResult<T> = Result<T, types::ErrorCode>;
+
+/// The error type used for HTTP operations within the WASI environment.
+///
+/// This enum provides a unified representation of all errors that can occur
+/// during HTTP request or response handling, whether they originate from
+/// WASI-level error codes, dynamic runtime failures, or full HTTP responses
+/// returned as error results.
+///
+/// # Variants
+///
+/// - [`Error::ErrorCode`]: Wraps a low-level [`wasip3::http::types::ErrorCode`]
+///   reported by the WASI HTTP runtime (e.g. connection errors, protocol errors).
+///
+/// - [`Error::Other`]: Represents an arbitrary dynamic error implementing
+///   [`std::error::Error`]. This allows integration with external libraries or
+///   application-specific failure types.
+///
+/// - [`Error::Response`]: Contains a full [`wasip3::http::types::Response`]
+///   representing an HTTP-level error (for example, a `4xx` or `5xx` response
+///   that should be treated as an error condition).
+///
+/// # See also
+/// - [`wasip3::http::types::ErrorCode`]: Standard WASI HTTP error codes.
+/// - [`wasip3::http::types::Response`]: Used when an error represents an HTTP response body.
+#[derive(Debug)]
+pub enum Error {
+    /// A low-level WASI HTTP error code.
+    ///
+    /// Wraps [`wasip3::http::types::ErrorCode`] to represent
+    /// transport-level or protocol-level failures.
+    ErrorCode(wasip3::http::types::ErrorCode),
+    /// A dynamic application or library error.
+    ///
+    /// Used for any runtime error that implements [`std::error::Error`],
+    /// allowing flexibility for different error sources.
+    Other(Box<dyn ::std::error::Error + Send + Sync>),
+    /// An HTTP response treated as an error.
+    ///
+    /// Contains a full [`wasip3::http::types::Response`], such as
+    /// a `404 Not Found` or `500 Internal Server Error`, when
+    /// the response itself represents an application-level failure.
+    Response(wasip3::http::types::Response),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Error::ErrorCode(e) => write!(f, "{e}"),
+            Error::Other(e) => write!(f, "{e}"),
+            Error::Response(e) => match http::StatusCode::from_u16(e.get_status_code()) {
+                Ok(status) => write!(f, "{status}"),
+                Err(e) => write!(f, "{e}"),
```

Comment:
I'm not sure `"{e}"` will show a good error message. Perhaps we hardcode it to `"invalid status code"`?